Added more non-standard tests

webbedspace opened the pull request:
- `Date.prototype.toLocaleFormat`
- error `columnNumber`
- `__lookupGetter__`
- String generics
- `for each...in`
- Iterator
- `__iterator__`
- Generators (JS 1.8)
- Array comprehensions
- Generator comprehensions

webbedspace committed:
Added a few more non-standard tests

* Date.prototype.toLocaleFormat
* error "columnNumber"
* `__lookupGetter__`

Also updated the `toSource` test a bit.

webbedspace committed:
Added more non-standard tests, 2

* String generics
* for each...in
* Iterator
* __iterator__

Also fixed Error#columnNumber and String#replace flags support values.

webbedspace committed:
Added non-standard array/generator c12s

c12s = comprehensions.

webbedspace committed:
Added non-standard old-style generator test

This tests for generators that use the JS 1.8 syntax - no asterisk by
the function keyword, and .next() returns an unwrapped value.

kangax commented on data-non-standard.js:
```diff
@@ -182,7 +182,15 @@ exports.tests = [
   name: '"toSource" method',
   exec: function () {
-    return 'toSource' in (function (){}) && 'toSource' in ({});
+    return Object   .prototype.hasOwnProperty('toSource')
```
Is there a reason for such precision? I think it's unnecessary.

webbedspace replied:
Well, I wanted to demonstrate that all of the prototypes possess their own unique 'toSource' methods. If I just ran `"toSource" in Function.prototype`, then that would return `true` if only Object.prototype possessed the toSource method that they all delegated to.

But, you're right - I guess since there's no standard for these, I should be permissive rather than rigorous for these tests. Need to get in the right mindframe.

webbedspace replied:
(Oh, well, if you mean "just that one line" rather than the entire set of checks in the test, then, well, OK, mostly I wrote it like that for visual consistency with the lines below - but it does seem leery.)

kangax commented on data-non-standard.js:
```diff
@@ -546,13 +554,51 @@ exports.tests = [
   name: 'Array generics',
   exec: function () {
-    return typeof Array.slice === 'function' && Array.slice('123').length === 3;
+    return typeof Array.slice === 'function' && Array.slice('abc').length === 3;
```
This is to avoid some sort of false positives?

webbedspace replied:
Mostly it's to make it 1% more readable (the first time I misread '123' as a number literal, interpreting it as some kind of nonsensical slice index).

kangax commented on data-non-standard.js:
```diff
@@ -585,6 +631,84 @@ exports.tests = [
 {
+  name: 'String generics',
+  exec: function () {
+    return typeof String.slice === 'function' && String.slice(123, 1) === "23";
```
There's also String generics? Nice.

kangax commented on data-non-standard.js:
```diff
@@ -663,6 +787,47 @@ exports.tests = [
   }
 },
 {
+  name: '"for each..in" loops',
```
Let's add a link to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for_each...in ?

kangax commented on data-non-standard.js:
```diff
@@ -704,6 +869,215 @@ exports.tests = [
+  exec:[{
+      type: 'application/javascript;version=1.8',
+      script: function () {
+	  console.log(JSON.stringify(__script_executed));
```
This is not needed, right?

webbedspace replied:
Whoops, sorry.

kangax commented on data-non-standard.js:
```diff
@@ -704,6 +869,215 @@ exports.tests = [
+      }
+  },{
+      script: function () {		
+	  console.log(JSON.stringify(__script_executed));
```
Also here

kangax commented on data-non-standard.js:
```diff
@@ -1033,6 +1407,41 @@ exports.tests = [
 {
+  name: 'Date.prototype.toLocaleFormat',
+  link: 'https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleFormat',
+  exec: function () { return typeof Date.prototype.toLocaleFormat == 'function' },
```
We should probably use `===` for consistency but not a big deal

kangax commented on data-non-standard.js:
```diff
@@ -1203,6 +1612,41 @@ exports.tests = [
   }
 },
 {
+  name: 'error "columnNumber"',
```
Let's add a link to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/columnNumber

kangax commented on data-non-standard.js:
```diff
@@ -1203,6 +1612,41 @@ exports.tests = [
+    firefox5: false,
+    firefox6: false,
+    firefox7: false,
+    firefox12: false,
```
Hm, I'm surprised it's false in FF12 and below.

kangax replied:
But you're right. Looks like introduced in FF17

webbedspace committed:
Cleanup + added more links

Removed some accidental console.log lines (groan), addressed a few
consistency/readability issues with the tests, and added more MDN links
for various tests.

I considered loosening some of the checks on the toSource method test
(such as avoid testing less useful methods like Number#toSource or
Error#toSource), but decided to leave it for now.

webbedspace commented:
Hope that commit is all good.

BTW, forgot to mention that this closes #151.

kangax commented:
I would still want to remove those extra `toSource` checks, but it's fine for now. Thanks for doing all this!
